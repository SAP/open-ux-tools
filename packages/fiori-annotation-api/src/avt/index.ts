export { convertAnnotationFile, AnnotationListWithOrigins } from './annotations';
export { convertMetadataToAvtSchema } from './metadata';
export {
    convertAnnotationToInternal,
    convertApplyToInternal,
    convertCollectionElement,
    convertCollectionToInternal,
    convertExpressionToInternal,
    convertPrimitiveValueToInternal,
    convertPropertyValueToInternal,
    convertRecordToInternal,
    convertTargetAnnotationsToInternal
} from './to-internal';
export { convertPointerInAnnotationToInternal } from './pointer';
export { resolvePath, isAnnotation, isAnnotationList, isCollection, isRecord } from './utils';
export { AVTNode } from './types';
export { getAvtNodeFromPointer, findAnnotation, findAnnotationByReference } from './find';
export { isExpression } from './expressions';
