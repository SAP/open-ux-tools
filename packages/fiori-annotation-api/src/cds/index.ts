export * from './service';
export { CDSAnnotationServiceAdapter } from './adapter';
