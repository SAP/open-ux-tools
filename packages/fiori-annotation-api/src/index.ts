export { FioriAnnotationService, FioriAnnotationServiceOptions } from './fiori-service';
export * from './protected';
export * from './types/change';
export { findAnnotation, getAvtNodeFromPointer } from './avt';
