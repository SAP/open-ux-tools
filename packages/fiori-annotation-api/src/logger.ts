import { ToolsLogger } from '@sap-ux/logger';

export const logger = new ToolsLogger({ logPrefix: '@sap-ux/fiori-annotation-api' });
