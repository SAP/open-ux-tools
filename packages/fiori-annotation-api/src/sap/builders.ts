import type { ElementChild, Element, Range } from '@sap-ux/odata-annotation-core-types';
import { Edm, createElementNode, createAttributeNode, Location } from '@sap-ux/odata-annotation-core-types';

import type { ValueWithOrigin } from './types';

function createComplexElement(elementName: Edm, nameAttribute: Edm, name: string, content: ElementChild[]): Element {
    return createElementNode({
        name: elementName,
        attributes: {
            [nameAttribute]: createAttributeNode(nameAttribute, name)
        },
        content
    });
}

function createPrimitiveElement(
    elementName: Edm,
    nameAttribute: Edm,
    name: string,
    valueType: Edm,
    value: string
): Element {
    return createElementNode({
        name: elementName,
        attributes: {
            [nameAttribute]: createAttributeNode(nameAttribute, name),
            [valueType]: createAttributeNode(valueType, value)
        }
    });
}

/**
 * Factory function for PropertyValue node with a primitive (attribute) value.
 *
 * @param name - Name of the property.
 * @param valueType - Type of the property.
 * @param value - Value of the property.
 * @returns PropertyValue node.
 */
export function createPrimitiveRecordProperty(name: string, valueType: Edm, value: string): Element {
    return createPrimitiveElement(Edm.PropertyValue, Edm.Property, name, valueType, value);
}

/**
 * Factory function for PropertyValue node with a complex (child element) value.
 *
 * @param name - Name of the property.
 * @param value - Value of the property.
 * @returns PropertyValue node.
 */
export function createComplexRecordProperty(name: string, value: Element): Element {
    return createComplexElement(Edm.PropertyValue, Edm.Property, name, [value]);
}

/**
 * Factory function for Annotation node with a primitive (attribute) value.
 *
 * @param termName - Annotation term.
 * @param valueType - Type of the annotation value.
 * @param value - Value of the annotation.
 * @returns Annotation node.
 */
export function createPrimitiveAnnotation(termName: string, valueType: Edm, value: string): Element {
    return createPrimitiveElement(Edm.Annotation, Edm.Term, termName, valueType, value);
}

/**
 * Factory function for Annotation node with a primitive (child element) value.
 *
 * @param termName - Annotation term.
 * @param value - Value of the annotation.
 * @returns Annotation node.
 */
export function createComplexAnnotation(termName: string, value: ElementChild): Element {
    return createElementNode({
        name: Edm.Annotation,
        attributes: {
            [Edm.Term]: createAttributeNode(Edm.Term, termName)
        },
        content: [value]
    });
}

/**
 * Factory function for Record node.
 *
 * @param properties - Record properties.
 * @param type - Record type.
 * @returns Record node.
 */
export function createRecord(properties: Element[], type?: string): Element {
    const node = createElementNode({
        name: Edm.Record,
        content: properties
    });
    if (type !== undefined) {
        node.attributes[Edm.Type] = createAttributeNode(Edm.Type, type);
    }
    return node;
}

/**
 * Factory function for value with origin.
 *
 * @param value - Value.
 * @param uri - File URI where the value is defined.
 * @param range - Range of the value in the file.
 * @returns ValueWithOrigin node.
 */
export function createValue<T>(value: T, uri?: string, range?: Range): ValueWithOrigin<T> {
    const result: ValueWithOrigin<T> = {
        value
    };

    if (uri && range) {
        result.location = Location.create(uri, range);
    }

    return result;
}
