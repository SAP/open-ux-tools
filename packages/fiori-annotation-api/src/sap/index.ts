export { SAPAnnotationConverter, convertTargets } from './converter';
