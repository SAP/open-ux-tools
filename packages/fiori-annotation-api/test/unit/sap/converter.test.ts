import { SAPAnnotationConverter } from '../../../src/sap/converter';
import { createValue } from '../../../src/sap/builders';
import { logger } from '../../../src/logger';

describe('SAP Annotation Converter', () => {
    const converter = new SAPAnnotationConverter();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    describe('UI LineItem', () => {
        test('empty', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.LineItem',
                    items: []
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
        test('with a data field', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.LineItem',
                    items: [
                        {
                            type: 'UI.DataField',
                            value: createValue('name')
                        }
                    ]
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
        test('multiple items with label', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.LineItem',
                    items: [
                        {
                            type: 'UI.DataField',
                            value: createValue('name')
                        },
                        {
                            type: 'UI.DataField',
                            value: createValue('category'),
                            label: createValue('Category')
                        }
                    ]
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
    });

    describe('UI FieldGroup', () => {
        test('empty', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.FieldGroup',
                    data: []
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
        test('with multiple data fields', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.FieldGroup',
                    data: [
                        {
                            type: 'UI.DataField',
                            value: createValue('name')
                        },
                        {
                            type: 'UI.DataField',
                            value: createValue('category')
                        }
                    ]
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
    });

    describe('UI Facets', () => {
        test('no field group', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.Facets',
                    facets: []
                }
            ]);
            expect(logger.warn).toHaveBeenCalledWith(
                'Could not find a property to which attach Facets annotation for entity "Service.Entity"'
            );
            expect(result).toMatchSnapshot();
        });

        test('empty', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    qualifier: createValue('general'),
                    term: 'UI.FieldGroup',
                    data: [
                        {
                            type: 'UI.DataField',
                            value: createValue('name')
                        }
                    ]
                },
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.Facets',
                    facets: []
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });

        test('reference facet to field group', () => {
            const result = converter.convertAnnotations([
                {
                    target: createValue('Service.Entity'),
                    qualifier: createValue('general'),
                    term: 'UI.FieldGroup',
                    data: [
                        {
                            type: 'UI.DataField',
                            value: createValue('name')
                        }
                    ]
                },
                {
                    target: createValue('Service.Entity'),
                    term: 'UI.Facets',
                    facets: [
                        {
                            type: 'UI.ReferenceFacet',
                            id: createValue('FacetID'),
                            target: createValue('general')
                        },
                        {
                            type: 'UI.ReferenceFacet',
                            id: createValue('FacetID2'),
                            target: createValue('general2'),
                            label: createValue('General 2')
                        }
                    ]
                }
            ]);
            expect(logger.warn).toHaveBeenCalledTimes(0);
            expect(result).toMatchSnapshot();
        });
    });
});
