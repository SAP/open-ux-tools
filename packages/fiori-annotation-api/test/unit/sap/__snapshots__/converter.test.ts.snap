// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SAP Annotation Converter UI Facets empty 1`] = `
Array [
  Object {
    "name": "Service.Entity/name",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.fieldGroup",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "qualifier",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "general",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.facet",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;

exports[`SAP Annotation Converter UI Facets no field group 1`] = `Array []`;

exports[`SAP Annotation Converter UI Facets reference facet to field group 1`] = `
Array [
  Object {
    "name": "Service.Entity/name",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.fieldGroup",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "qualifier",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "general",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.facet",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "id",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "FacetID",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "STANDARD",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "purpose",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "FIELDGROUP_REFERENCE",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "type",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "targetQualifier",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "general",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "id",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "FacetID2",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "STANDARD",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "purpose",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "FIELDGROUP_REFERENCE",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "type",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "targetQualifier",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "general2",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "20",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "label",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "General 2",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;

exports[`SAP Annotation Converter UI FieldGroup empty 1`] = `Array []`;

exports[`SAP Annotation Converter UI FieldGroup with multiple data fields 1`] = `
Array [
  Object {
    "name": "Service.Entity/name",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.fieldGroup",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
  Object {
    "name": "Service.Entity/category",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.fieldGroup",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "20",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;

exports[`SAP Annotation Converter UI LineItem empty 1`] = `Array []`;

exports[`SAP Annotation Converter UI LineItem multiple items with label 1`] = `
Array [
  Object {
    "name": "Service.Entity/name",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.lineItem",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "HIGH",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "importance",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
  Object {
    "name": "Service.Entity/category",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.lineItem",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "20",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "HIGH",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "importance",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
      Object {
        "attributes": Object {
          "String": Object {
            "name": "String",
            "type": "attribute",
            "value": "Category",
          },
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "EndUserText.label",
          },
        },
        "content": Array [],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;

exports[`SAP Annotation Converter UI LineItem with a data field 1`] = `
Array [
  Object {
    "name": "Service.Entity/name",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.lineItem",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "HIGH",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "importance",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;

exports[`SAP Annotation Converter combinations multiple labels for the same property 1`] = `
Array [
  Object {
    "name": "Service.Entity/category",
    "nameRange": undefined,
    "range": undefined,
    "terms": Array [
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.lineItem",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "EnumMember": Object {
                        "name": "EnumMember",
                        "type": "attribute",
                        "value": "HIGH",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "importance",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
      Object {
        "attributes": Object {
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "UI.fieldGroup",
          },
        },
        "content": Array [
          Object {
            "attributes": Object {},
            "content": Array [
              Object {
                "attributes": Object {},
                "content": Array [
                  Object {
                    "attributes": Object {
                      "Int": Object {
                        "name": "Int",
                        "type": "attribute",
                        "value": "10",
                      },
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "position",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                  Object {
                    "attributes": Object {
                      "Property": Object {
                        "name": "Property",
                        "type": "attribute",
                        "value": "qualifier",
                      },
                      "String": Object {
                        "name": "String",
                        "type": "attribute",
                        "value": "general",
                      },
                    },
                    "content": Array [],
                    "name": "PropertyValue",
                    "type": "element",
                  },
                ],
                "name": "Record",
                "type": "element",
              },
            ],
            "name": "Collection",
            "type": "element",
          },
        ],
        "name": "Annotation",
        "type": "element",
      },
      Object {
        "attributes": Object {
          "String": Object {
            "name": "String",
            "type": "attribute",
            "value": "Category From LineItem",
          },
          "Term": Object {
            "name": "Term",
            "type": "attribute",
            "value": "EndUserText.label",
          },
        },
        "content": Array [],
        "name": "Annotation",
        "type": "element",
      },
    ],
    "termsRange": undefined,
    "type": "target",
  },
]
`;
